Mathlib/Algebra/Module/Equiv.lean

```diff
@@ -0,0 +1,913 @@
+
+-- mathport name: «expr ≃ₛₗ[ ] »
+notation:50 M " ≃ₛₗ[" σ "] " M₂ => LinearEquiv σ M M₂
+
+-- mathport name: «expr ≃ₗ[ ] »
+notation:50 M " ≃ₗ[" R "] " M₂ => LinearEquiv (RingHom.id R) M M₂
+
+-- mathport name: «expr ≃ₗ⋆[ ] »
+notation:50 M " ≃ₗ⋆[" R "] " M₂ => LinearEquiv (starRingEnd R) M M₂
+
+/-- `semilinear_equiv_class F σ M M₂` asserts `F` is a type of bundled `σ`-semilinear equivs
+`M → M₂`.
+
+See also `linear_equiv_class F R M M₂` for the case where `σ` is the identity map on `R`.
+
+A map `f` between an `R`-module and an `S`-module over a ring homomorphism `σ : R →+* S`
+is semilinear if it satisfies the two properties `f (x + y) = f x + f y` and
+`f (c • x) = (σ c) • f x`. -/
+class SemilinearEquivClass (F : Type _) {R S : outParam (Type _)} [Semiring R] [Semiring S]
+  (σ : outParam <| R →+* S) {σ' : outParam <| S →+* R} [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+  (M M₂ : outParam (Type _)) [AddCommMonoid M] [AddCommMonoid M₂] [Module R M] [Module S M₂] extends
+  AddEquivClass F M M₂ where
+  map_smulₛₗ : ∀ (f : F) (r : R) (x : M), f (r • x) = σ r • f x
+#align semilinear_equiv_class SemilinearEquivClass
+
+-- `R, S, σ, σ'` become metavars, but it's OK since they are outparams.
+--Porting note: TODO attribute [nolint dangerous_instance] SemilinearEquivClass.toAddEquivClass
+
+/-- `linear_equiv_class F R M M₂` asserts `F` is a type of bundled `R`-linear equivs `M → M₂`.
+This is an abbreviation for `semilinear_equiv_class F (ring_hom.id R) M M₂`.
+-/
+abbrev LinearEquivClass (F : Type _) (R M M₂ : outParam (Type _)) [Semiring R] [AddCommMonoid M]
+    [AddCommMonoid M₂] [Module R M] [Module R M₂] :=
+  SemilinearEquivClass F (RingHom.id R) M M₂
+#align linear_equiv_class LinearEquivClass
+
+end
+
+namespace SemilinearEquivClass
+
+variable (F : Type _) [Semiring R] [Semiring S]
+
+variable [AddCommMonoid M] [AddCommMonoid M₁] [AddCommMonoid M₂]
+
+variable [Module R M] [Module S M₂] {σ : R →+* S} {σ' : S →+* R}
+
+-- `σ'` becomes a metavariable, but it's OK since it's an outparam
+--Porting note: TODO @[nolint dangerous_instance]
+instance (priority := 100) [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+    [s : SemilinearEquivClass F σ M M₂] : SemilinearMapClass F σ M M₂ :=
+  { s with
+    coe := (s.coe : F → M → M₂)
+    coe_injective' := @FunLike.coe_injective F _ _ _ }
+
+end SemilinearEquivClass
+
+namespace LinearEquiv
+
+section AddCommMonoid
+
+variable {M₄ : Type _}
+
+variable [Semiring R] [Semiring S]
+
+section
+
+variable [AddCommMonoid M] [AddCommMonoid M₁] [AddCommMonoid M₂]
+
+variable [Module R M] [Module S M₂] {σ : R →+* S} {σ' : S →+* R}
+
+variable [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+
+--Porting note: TODO include R
+
+--Porting note: TODO include σ'
+
+instance : Coe (M ≃ₛₗ[σ] M₂) (M →ₛₗ[σ] M₂) :=
+  ⟨toLinearMap⟩
+
+-- see Note [function coercion]
+instance : CoeFun (M ≃ₛₗ[σ] M₂) fun _ => M → M₂ :=
+  ⟨(·.toFun)⟩ --Porting note: TODO why does `⟨toFun⟩` not suffice?
```

Comment:
Instances like these of `CoeFun` when there's already a `FunLike` instance should be deleted in the port, since we want `FunLike.coe` as the normal form.